### PR TITLE
MSGN: avoid 'Scanline corrupt' report for HRV with or without RSS

### PR DESCRIPTION
**Pull request by g8sqh**
Using archive data from https://data.eumetsat.int/data/map/EO:EUM:DAT:MSG:HRSEVIRI, whole earth images are prefixed MSG4-, RSS (every 5 minutes, only Europe) are MSG3-

To extract the HRV image, filename must be given as HRV:

Commit 84e31284211509314c84a41d78bd12962b9b170d allowed extraction of HRV images from whole-disk (MSG4-) files

RSS HRV (MSG3-) reports 'Scanline corrupt' - in fact the extracted image would be OK

This commit works around poGDS->m_bHRVDealWithSplit being set false for HRV RSS images (in msgndataset.cpp).

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: frmts/msgn/msgndataset.cpp [hunk 1]
```diff
@@ -220,8 +220,18 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
...
+				      poGDS->msg_reader_core->get_line_start())) != (unsigned int)i_nBlockYOff )
     {
+        fprintf(stderr, "Deal with split %s\n", poGDS->m_bHRVDealWithSplit ? "yes" : "no" );
+        fprintf(stderr, "nread = %lu, data_len %d, linenum %d, start %d, offset %d\n",
+		(long unsigned int) nread, // Mingw_w64 otherwise wants %llu - MSG read will never exceed 32 bits
+               data_length,
+               (p->lineNumberInVisirGrid),
+               poGDS->msg_reader_core->get_line_start(),
+               i_nBlockYOff);
+
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:234]: can you remove those fprintf() statement, or change them to CPLDebug() ones ?

**Review by rouault — COMMENTED** [review]
Please remove tabulation characters and replace them with space indentation

File: frmts/msgn/msgndataset.cpp [hunk 2]
```diff
@@ -61,16 +67,16 @@ class MSGNDataset final: public GDALDataset
 
     Msg_reader_core*    msg_reader_core;
     open_mode_type      m_open_mode = MODE_VISIR;
-    bool                m_bHRVDealWithSplit = false;
+    hrv_mode_type       m_bHRVDealWithSplit = NOT_HRV;
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:70]: The 'b' means boolean, hence changing to 'e' to mean enumeration:

```suggestion
    hrv_mode_type       m_eHRVDealWithSplit = NOT_HRV;
```

File: frmts/msgn/msgndataset.cpp [hunk 3]
```diff
@@ -209,7 +215,7 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
     SUB_VISIRLINE* p = (SUB_VISIRLINE*) pszRecord;
     to_native(*p);
 
-    if (p->lineValidity != 1 || poGDS->m_bHRVDealWithSplit) {
+    if (p->lineValidity != 1  || poGDS->m_bHRVDealWithSplit) {  // Split lines are not full width, so NODATA all first
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:218]:
```suggestion
    if (p->lineValidity != 1  || poGDS->m_eHRVDealWithSplit != NOT_HRV) {  // Split lines are not full width, so NODATA all first
```

File: frmts/msgn/msgndataset.cpp [hunk 4]
```diff
@@ -220,8 +226,19 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
     }
 
     if ( nread != data_length ||
-        (p->lineNumberInVisirGrid - poGDS->msg_reader_core->get_line_start()) != (unsigned int)i_nBlockYOff )
+         (p->lineNumberInVisirGrid - ((open_mode == MODE_HRV && poGDS->m_bHRVDealWithSplit == HRV_RSS) ?
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:229]:
```suggestion
         (p->lineNumberInVisirGrid - ((open_mode == MODE_HRV && poGDS->m_eHRVDealWithSplit == HRV_RSS) ?
```

File: frmts/msgn/msgndataset.cpp [hunk 5]
```diff
@@ -220,8 +226,19 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
...
+                                      (3 * poGDS->msg_reader_core->get_line_start()) - 2 :
+                                      poGDS->msg_reader_core->get_line_start())) != (unsigned int)i_nBlockYOff )
     {
+        CPLDebug("MSGN", "Deal with split %s", poGDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:233]:
```suggestion
        CPLDebug("MSGN", "Deal with split %s", poGDS->m_eHRVDealWithSplit == HRV_RSS ? "RSS" :
```

File: frmts/msgn/msgndataset.cpp [hunk 6]
```diff
@@ -220,8 +226,19 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
...
+                 (poGDS->m_bHRVDealWithSplit ? "whole" : "no") );
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:234]:
```suggestion
                 (poGDS->m_eHRVDealWithSplit ? "whole" : "no") );
```

File: frmts/msgn/msgndataset.cpp [hunk 7]
```diff
@@ -472,17 +494,40 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
             idr.plannedCoverage_hrv.upperWestColumnPlanned <= poDS->nRasterXSize * 3 )
         {
             poDS->nRasterXSize *= 3;
-            poDS->m_bHRVDealWithSplit = true;
+            poDS->m_bHRVDealWithSplit = HRV_WHOLE_DISK;
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:497]:
```suggestion
            poDS->m_eHRVDealWithSplit = HRV_WHOLE_DISK;
```

File: frmts/msgn/msgndataset.cpp [hunk 8]
```diff
@@ -472,17 +494,40 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
+                 idr.plannedCoverage_hrv.lowerSouthLinePlanned > 1 &&
+                 idr.plannedCoverage_hrv.lowerSouthLinePlanned < idr.referencegrid_hrv.numberOfLines &&
+                 idr.plannedCoverage_hrv.lowerEastColumnPlanned >= 1 &&
+                 idr.plannedCoverage_hrv.lowerWestColumnPlanned <= poDS->nRasterXSize * 3 &&
+                 // full height
+                 idr.plannedCoverage_hrv.lowerNorthLinePlanned == idr.plannedCoverage_hrv.lowerSouthLinePlanned + poDS->nRasterYSize -1
+                 )
+        {
+            poDS->nRasterXSize *= 3;
+            poDS->m_bHRVDealWithSplit = HRV_RSS;
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:518]:
```suggestion
            poDS->m_eHRVDealWithSplit = HRV_RSS;
```

File: frmts/msgn/msgndataset.cpp [hunk 9]
```diff
@@ -472,17 +494,40 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
+            poDS->m_nHRVLowerShiftX = idr.plannedCoverage_hrv.lowerEastColumnPlanned - 1;
+            poDS->m_nHRVUpperShiftX = 0; // should never be used - could set huge to trigger SEGV
+        }
         else
         {
             // In practice, the number of columns of HRV seems to be 1.5x the one
             // of VISIR bands, but derive it from the paquet header
             poDS->nRasterXSize = nRawHRVColumns;
         }
+        CPLDebug("MSGN", "Deal with split %s", poDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:529]:
```suggestion
        CPLDebug("MSGN", "Deal with split %s", poDS->m_eHRVDealWithSplit == HRV_RSS ? "RSS" :
```

File: frmts/msgn/msgndataset.cpp [hunk 10]
```diff
@@ -472,17 +494,40 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
+                 (poDS->m_bHRVDealWithSplit ? "whole" : "no"));
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:530]:
```suggestion
                 (poDS->m_eHRVDealWithSplit ? "whole" : "no"));
```

**Comment by g8sqh** [comment]
Dear Even

Thank you for improving what I wrote - I wasn't aware of the naming convention.

As you can tell, this is my first time contributing to GDAL - git is familiar to me, but all of the github machinery is new to me. Tell me if there are other things I need to do to to finish this pull request.

Thanks again

--David


On Sun, 05 Jun 2022 02:07:35 -0700
Even Rouault ***@***.***> wrote:

> @rouault commented on this pull request.
> 
> 
> 
> > @@ -61,16 +67,16 @@ class MSGNDataset final: public GDALDataset
>  
>      Msg_reader_core*    msg_reader_core;
>      open_mode_type      m_open_mode = MODE_VISIR;
> -    bool                m_bHRVDealWithSplit = false;
> +    hrv_mode_type       m_bHRVDealWithSplit = NOT_HRV;
> 
> The 'b' means boolean, hence changing to 'e' to mean enumeration:
> 
> ```suggestion
>     hrv_mode_type       m_eHRVDealWithSplit = NOT_HRV;
> ```
> 
> > @@ -209,7 +215,7 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
>      SUB_VISIRLINE* p = (SUB_VISIRLINE*) pszRecord;
>      to_native(*p);
>  
> -    if (p->lineValidity != 1 || poGDS->m_bHRVDealWithSplit) {
> +    if (p->lineValidity != 1  || poGDS->m_bHRVDealWithSplit) {  // Split lines are not full width, so NODATA all first
> 
> ```suggestion
>     if (p->lineValidity != 1  || poGDS->m_eHRVDealWithSplit != NOT_HRV) {  // Split lines are not full width, so NODATA all first
> ```
> 
> > @@ -220,8 +226,19 @@ CPLErr MSGNRasterBand::IReadBlock( CPL_UNUSED int nBlockXOff,
>      }
>  
>      if ( nread != data_length ||
> -        (p->lineNumberInVisirGrid - poGDS->msg_reader_core->get_line_start()) != (unsigned int)i_nBlockYOff )
> +         (p->lineNumberInVisirGrid - ((open_mode == MODE_HRV && poGDS->m_bHRVDealWithSplit == HRV_RSS) ?
> 
> ```suggestion
>          (p->lineNumberInVisirGrid - ((open_mode == MODE_HRV && poGDS->m_eHRVDealWithSplit == HRV_RSS) ?
> ```
> 
> >      {
> +        CPLDebug("MSGN", "Deal with split %s", poGDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
> 
> ```suggestion
>         CPLDebug("MSGN", "Deal with split %s", poGDS->m_eHRVDealWithSplit == HRV_RSS ? "RSS" :
> ```
> 
> >      {
> +        CPLDebug("MSGN", "Deal with split %s", poGDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
> +                 (poGDS->m_bHRVDealWithSplit ? "whole" : "no") );
> 
> ```suggestion
>                  (poGDS->m_eHRVDealWithSplit ? "whole" : "no") );
> ```
> 
> > @@ -472,17 +494,40 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
>              idr.plannedCoverage_hrv.upperWestColumnPlanned <= poDS->nRasterXSize * 3 )
>          {
>              poDS->nRasterXSize *= 3;
> -            poDS->m_bHRVDealWithSplit = true;
> +            poDS->m_bHRVDealWithSplit = HRV_WHOLE_DISK;
> 
> ```suggestion
>             poDS->m_eHRVDealWithSplit = HRV_WHOLE_DISK;
> ```
> 
> > +                 idr.plannedCoverage_hrv.upperSouthLinePlanned == 0 &&
> +                 idr.plannedCoverage_hrv.upperWestColumnPlanned == 0 &&
> +                 idr.plannedCoverage_hrv.upperEastColumnPlanned == 0 &&  // RSS only uses the lower section
> +                 idr.plannedCoverage_hrv.lowerNorthLinePlanned == idr.referencegrid_hrv.numberOfLines && // start at max N
> +                 // full expected width
> +                 idr.plannedCoverage_hrv.lowerWestColumnPlanned == idr.plannedCoverage_hrv.lowerEastColumnPlanned + nRawHRVColumns - 1 &&
> +                 idr.plannedCoverage_hrv.lowerSouthLinePlanned > 1 &&
> +                 idr.plannedCoverage_hrv.lowerSouthLinePlanned < idr.referencegrid_hrv.numberOfLines &&
> +                 idr.plannedCoverage_hrv.lowerEastColumnPlanned >= 1 &&
> +                 idr.plannedCoverage_hrv.lowerWestColumnPlanned <= poDS->nRasterXSize * 3 &&
> +                 // full height
> +                 idr.plannedCoverage_hrv.lowerNorthLinePlanned == idr.plannedCoverage_hrv.lowerSouthLinePlanned + poDS->nRasterYSize -1
> +                 )
> +        {
> +            poDS->nRasterXSize *= 3;
> +            poDS->m_bHRVDealWithSplit = HRV_RSS;
> 
> ```suggestion
>             poDS->m_eHRVDealWithSplit = HRV_RSS;
> ```
> 
> >          else
>          {
>              // In practice, the number of columns of HRV seems to be 1.5x the one
>              // of VISIR bands, but derive it from the paquet header
>              poDS->nRasterXSize = nRawHRVColumns;
>          }
> +        CPLDebug("MSGN", "Deal with split %s", poDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
> 
> ```suggestion
>         CPLDebug("MSGN", "Deal with split %s", poDS->m_eHRVDealWithSplit == HRV_RSS ? "RSS" :
> ```
> 
> >          else
>          {
>              // In practice, the number of columns of HRV seems to be 1.5x the one
>              // of VISIR bands, but derive it from the paquet header
>              poDS->nRasterXSize = nRawHRVColumns;
>          }
> +        CPLDebug("MSGN", "Deal with split %s", poDS->m_bHRVDealWithSplit == HRV_RSS ? "RSS" :
> +                 (poDS->m_bHRVDealWithSplit ? "whole" : "no"));
> 
> ```suggestion
>                  (poDS->m_eHRVDealWithSplit ? "whole" : "no"));
> ```
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/OSGeo/gdal/pull/5866#pullrequestreview-995988632
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>


-- 
g8sqh ***@***.***>

**Comment by rouault** [comment]
I've tested this, and I observe that within a RSS dataset, the georeferencing of the HRV and non-HRV band is consistent (when displaying in QGIS for example). But it is not consistent with MSGN datasets of whole-Earth images. There's a vertical shift of about a Earth radius. Some tweaking of the origin_y would be needed, or we'd probably need to declare the dataset height as being the whole Earth and tweak IReadBlock() to apply  a vertical shift

**Comment by g8sqh** [comment]
Ok, the vertical shift is interesting (and I hadn't spotted it). At present I can't get qgis to convert from the geostationary projection (EPSG 9001) to lat/long - when I set the project CRS to (say) WGS-84 it won't show the image. (And I need to raise a QGIS question - I can't see how to pass HRV:<file name> in - and so qgis for me won't directly extract HRV - it does see bands 1-11)

If qgis is able to read and convert EPSG 9001, then we need to keep the image location within the raster aligned with the projection data in the .nat file. gdalinfo on HRV:MSG4... looks strange - the origin is 5565, -5565km and the bottom right corner is (3*5565, -3*5565)  I think this may be your one earth radius offset.  I need to reread the nat file specification to understand this.

**Comment by rouault** [comment]
> I can't see how to pass HRV: in

if you manually specify HRV:full_path_name in Add Raster layer, that works.
But you can also use gdal_translate HRV:.... out.tif (or to a VRT file)

**Comment by rouault** [comment]
```
/home/travis/build/OSGeo/gdal/frmts/msgn/msgndataset.cpp: In static member function ‘static GDALDataset* MSGNDataset::Open(GDALOpenInfo*)’:
/home/travis/build/OSGeo/gdal/frmts/msgn/msgndataset.cpp:597:11: error: declaration of ‘i’ shadows a previous local [-Werror=shadow]
       int i;
           ^
/home/travis/build/OSGeo/gdal/frmts/msgn/msgndataset.cpp:539:18: note: shadowed declaration is here
     unsigned int i;
                  ^
```

**Comment by g8sqh** [comment]
I believe this now works for MSG4, RSS from MSG3 and IODC from MSG2 - all for normal and HRV channels.
I've hand-tweaked small constants to get decent registration to country boundaries - but no guarantee on lat/long from this

In qgis this will show the circular originals (with no CRS set), and also reproject onto a normal CRS (e.g. EPSG 3857)

File: frmts/msgn/CMakeLists.txt [hunk 11]
```diff
@@ -8,7 +8,7 @@ add_gdal_driver(
           PLUGIN_CAPABLE
           NO_DEPS)
 gdal_standard_includes(gdal_MSGN)
-target_compile_definitions(gdal_MSGN PRIVATE -DGDAL_SUPPORT)
+target_compile_definitions(gdal_MSGN PRIVATE )
```
Comment (rouault) [inline frmts/msgn/CMakeLists.txt:11]: -DGDAL_SUPPORT should be kept

File: frmts/msgn/msg_reader_core.cpp [hunk 12]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
 {
     memset(&_main_header, 0, sizeof(_main_header));
     memset(&_sec_header, 0, sizeof(_sec_header));
-
+    
     SecondaryProdHeaderInit(&_sec_header);
     for (size_t i=0; i < MSG_NUM_CHANNELS; ++i) {
-      _calibration[i].cal_slope = 0.0;
-      _calibration[i].cal_offset = 0.0;
+    _calibration[i].cal_slope = 0.0;
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:177]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 13]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
+    
 #ifdef DEBUG
     // print out all the fields in the header
     PH_DATA* hd = (PH_DATA*)&_main_header;
-    for (int i=0; i < 6; i++) {
-        to_string(*hd);
-        printf("[%02d] %s %s", i, hd->name, hd->value);/*ok*/
-        hd++;
+    for (i=0; i < 6; i++) {
+    to_string(*hd);
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:199]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 14]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
+    printf("[%02u] %s %s", i, hd->name, hd->value);/*ok*/
+    hd++;
     }
-    PH_DATA_ID* hdi = (PH_DATA_ID*)&_main_header.dataSetIdentification;
-
+    
     for (i=0; i < 5; i++) {
-        printf("%s %s %s", hdi->name, hdi->size, hdi->address);/*ok*/
-        hdi++;
+    printf("%s %s %s", hdi->name, hdi->size, hdi->address);/*ok*/
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:205]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 15]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
+    hdi++;
     }
     hd = (PH_DATA*)(&_main_header.totalFileSize);
-    for (int i=0; i < 19; i++) {
-        to_string(*hd);
-        printf("[%02d] %s %s", i, hd->name, hd->value);/*ok*/
-        hd++;
+    for (i=0; i < 19; i++) {
+    to_string(*hd);
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:210]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 16]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
-        PH_DATA_ID* hdi = (PH_DATA_ID*)&_main_header.dataSetIdentification[i];
-        if (STARTS_WITH(hdi->name, "15Header")) {
-            sscanf(hdi->size, "%u", &_f_header_size);
-            sscanf(hdi->address, "%u", &_f_header_offset);
-        } else
-            if (STARTS_WITH(hdi->name, "15Data")) {
-            sscanf(hdi->size, "%u", &_f_data_size);
-            sscanf(hdi->address, "%u", &_f_data_offset);
-        }
+    hdi = (PH_DATA_ID*)&_main_header.dataSetIdentification[i];
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:219]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 17]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
-    for (unsigned int i=0; i < MSG_NUM_CHANNELS; i++) {
-        if (_calibration[i].cal_slope < 0 || _calibration[i].cal_slope > 0.4)
-        {
-            printf("Warning: calibration slope (%f) out of nominal range. MSG reader probably broken\n", _calibration[i].cal_slope);/*ok*/
-        }
-        if (_calibration[i].cal_offset > 0 || _calibration[i].cal_offset < -20) {
-            printf("Warning: calibration offset (%f) out of nominal range. MSG reader probably broken\n",/*ok*/ _calibration[i].cal_offset);
-        }
+    for (i=0; i < MSG_NUM_CHANNELS; i++) {
+    if (_calibration[i].cal_slope < 0 || _calibration[i].cal_slope > 0.4)
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:289]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 18]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
     _interline_spacing = 0;
     visir_line.channelId = 0;
-
+    
     int scanned_bands[MSG_NUM_CHANNELS];
     int band_count = 0;
     for (i=0; i < MSG_NUM_CHANNELS; i++) {
-        scanned_bands[i] = _bands[i];
-        band_count += _bands[i];
+    scanned_bands[i] = _bands[i];
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:356]: lack of indentation

File: frmts/msgn/msg_reader_core.cpp [hunk 19]
```diff
@@ -163,298 +171,366 @@ Msg_reader_core::Msg_reader_core( VSILFILE* fp ) :
...
+    band_count += _bands[i];
     }
-
+    
     do {
-        if( VSIFReadL(&gp_header, sizeof(GP_PK_HEADER), 1, fin) != 1 ||
+    if( VSIFReadL(&gp_header, sizeof(GP_PK_HEADER), 1, fin) != 1 ||
```
Comment (rouault) [inline frmts/msgn/msg_reader_core.cpp:361]: lack of indentation

**Review by rouault — COMMENTED** [review]
hard to review due to many wrong changes in indentation (I stopped at mentioning them...) or addition of lines with white spaces only.  Could you try starting from a fresh branch and fix your editor and apply only the relevant changes ?

**Comment by g8sqh** [comment]
I'll try again - I probably don't have the git skills to get this to work - may just keep as a patch set locally

**Pull request closed without merging**